### ISSUE TITLE
Partitioner's auto-AC misbehaves with mixed dtypes

### ISSUE THREAD

**Issue by lw** [pytorch/pytorch#144470]
### 🚀 The feature, motivation and pitch

When using the partitioner to do some automatic selective activation checkpointing, the default mode consists in estimating the runtime cost of the operators by computing the number of FLOPs they consume. Unfortunately, this is a bad proxy metric in cases where different operators are executed with different dtypes because, in a sense, not all FLOPs are equal.

Concretely, this occurs when using fp8 matmuls on H100 GPUs, because (in most/all current recipes) only the matmuls are converted to fp8, whereas the self-attention remains in bf16. Moreover, in some cases only _some_ matmuls get converted to fp8 (e.g., some layers, or some specific weights within layers).

If the partitioner just compares FLOPs without adjusting them by the time it takes to execute a FLOP in a given dtype this might lead to a suboptimal solution to the AC problem.

**Comment by lw** [comment]
CC @Chillee

**Comment by Chillee** [comment]
@lw Out of curiosity, is this a practical problem?

You can also do `torch._functorch.config.activation_memory_budget_runtime_estimator = "runtime"`, which will profile each op instead of analytically estimating it.

**Comment by lw** [comment]
I didn't run into this problem in practice, _as far as I'm aware_. It's not easy to introspect into the decisions of AutoAC hence I don't know if the policy it chose was suboptimal or not. Though we did have runs where we were mixing fp8 and bf16 and this problem could have emerged.

I'll consider using runtime estimations in the future. I guess this can also help when we have inefficient kernels (i.e., using FA2 on H100) which "take longer" per FLOP.

**Comment by lw** [comment]
How "robust" is the runtime estimation to distributed training? As in, if there's noise in the profiling, could different ranks end up doing different decisions and, if collectives are involved, could this cause a hang?

I guess, ideally, the ranks could do some kind of "vote" on which kernels they think are the fastest, and this could even reduce the profiling noise. Is something of this kind implemented?
